### PR TITLE
Interface Material

### PR DESCRIPTION
**Pull request by lindsayad**
Create an `InterfaceMaterial` class capable of accessing face and neighboring face variable values and material properties

### CODE REVIEW

**Comment by permcody** [comment]
Nice! - Oh goodie, looks like Rattlesnake gets broken again...

**Comment by YaqiWang** [comment]
Happy to see that!

**Comment by lindsayad** [comment]
Is anyone every gonna review this?

**Comment by permcody** [comment]
I looked over this and didn't see any show stoppers. My comment should have been a little bit more specific. Are you going to make the patch for Rattlesnake or do you need help here? Perhaps @YaqiWang can help push this one over the finish line?

File: test/src/materials/InterfaceValueMaterial.C [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+                             "Type of scalar output");
+  return params;
+}
+
+InterfaceValueMaterial::InterfaceValueMaterial(const InputParameters & parameters)
+  : InterfaceMaterial(parameters),
+    _mp_master_name(getParam<std::string>("mat_prop_master")),
+    _mp_slave_name(getParam<std::string>("mat_prop_slave")),
+    _mp_master(getMaterialPropertyByName<Real>(_mp_master_name)),
+    _mp_slave(getNeighborMaterialProperty<Real>(_mp_slave_name)),
```
Comment (YaqiWang) [inline test/src/materials/InterfaceValueMaterial.C]: Why this one do not have `ByName` in it?

File: test/src/materials/InterfaceValueMaterial.C [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+    _mp_var_out_base_name(getParam<std::string>("mat_prop_var_out_basename")),
+    _interface_value(
+        declareProperty<Real>(_mp_out_base_name + "_" + std::string(_interface_value_type))),
+    _interface_value_2(
+        declareProperty<Real>(_mp_var_out_base_name + "_" + std::string(_interface_value_type))),
+    _interface_value_old(
+        getMaterialPropertyOld<Real>(_mp_out_base_name + "_" + std::string(_interface_value_type))),
+    _interface_value_2_old(getMaterialPropertyOld<Real>(_mp_var_out_base_name + "_" +
+                                                        std::string(_interface_value_type))),
+    _jump(declareProperty<Real>(static_cast<std::string>(_interface_value_type) + "_jump"))
```
Comment (YaqiWang) [inline test/src/materials/InterfaceValueMaterial.C]: We have to have static_cast here, or std::string itself is not enough?

**Comment by lindsayad** [comment]
I will create a patch for Rattlesnake. Based on the error, it looks like it should be a very small fix. I'll make the fix now, and let you know when it's ready.

File: test/src/materials/InterfaceValueMaterial.C [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+{
+}
+
+void
+InterfaceValueMaterial::computeQpProperties()
+{
+  mooseAssert(_mp_master[_qp] == _var_master[_qp],
```
Comment (YaqiWang) [inline test/src/materials/InterfaceValueMaterial.C]: Not obvious to me why we need these two assertions here.

File: test/tests/interfacekernels/1d_interface/coupled_value_coupled_flux_with_jump_material.i [hunk 4]
```diff
@@ -0,0 +1,125 @@
+[Mesh]
+  type = GeneratedMesh
+  dim = 1
+  nx = 10
+  xmax = 2
+[]
+
+[MeshModifiers]
```
Comment (YaqiWang) [inline test/tests/interfacekernels/1d_interface/coupled_value_coupled_flux_with_jump_material.i]: I guess we are moving to the new mesh generation system.

File: test/tests/interfacekernels/1d_interface/coupled_value_coupled_flux_with_jump_material.i [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+  [../]
+[]
+
+[InterfaceKernels]
+  [./penalty_interface]
+    type = PenaltyInterfaceDiffusion
+    variable = u
+    neighbor_var = v
+    boundary = master0_interface
+    penalty = 1e6
```
Comment (YaqiWang) [inline test/tests/interfacekernels/1d_interface/coupled_value_coupled_flux_with_jump_material.i]: It makes more sense to me to change penalty to an interface material property. `jump_prop_name` makes this kernel too toylish.

File: test/tests/interfacekernels/1d_interface/tests [hunk 6]
```diff
@@ -18,6 +18,16 @@
...
     design = 'InterfaceKernels/index.md'
   [../]
+  [./interface_diffusion_penalty_with_jump_material]
+    type = 'Exodiff'
+    input = 'coupled_value_coupled_flux_with_jump_material.i'
+    exodiff = 'coupled_value_coupled_flux_with_jump_material_out.e'
+    design = 'InterfaceKernels/index.md'
+    issues = '#12066'
+    requirement = "The InterfaceKernel system shall use with interface material in 1D."
+                  "In this penalty implementation, jump is calculated by an interface material."
```
Comment (YaqiWang) [inline test/tests/interfacekernels/1d_interface/tests]: This line does not have to be in requirement.

File: test/tests/interfacekernels/2d_interface/tests [hunk 7]
```diff
@@ -20,6 +20,17 @@
...
 
+  [./test_jump_material]
+    type = 'Exodiff'
+    input = 'coupled_value_coupled_flux_with_jump_material.i'
+    exodiff = 'coupled_value_coupled_flux_with_jump_material_out.e'
+    mesh_mode = REPLICATED
+    issues = "#12066"
+    design = "InterfaceKernels/index.md"
+    requirement = "The InterfaceKernel system shall use with interface material in 2D."
+                  "In this penalty implementation, jump is calculated by an interface material."
```
Comment (YaqiWang) [inline test/tests/interfacekernels/2d_interface/tests]: Same.

File: test/tests/interfacekernels/3d_interface/coupled_value_coupled_flux_with_jump_material.i [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+  nx = 2
+  xmax = 2
+  ny = 2
+  ymax = 2
+  nz = 2
+  zmax = 2
+[]
+
+[MeshModifiers]
+  [./subdomain1]
```
Comment (YaqiWang) [inline test/tests/interfacekernels/3d_interface/coupled_value_coupled_flux_with_jump_material.i]: new mesh generation system.

File: test/tests/interfacekernels/3d_interface/tests [hunk 9]
```diff
@@ -18,4 +18,14 @@
...
   [../]
+  [./test_jump_material]
+    type = 'Exodiff'
+    input = 'coupled_value_coupled_flux_with_jump_material.i'
+    exodiff = 'coupled_value_coupled_flux_with_jump_material_out.e'
+    mesh_mode = REPLICATED
+    issues = "#12066"
+    design = "InterfaceKernels/index.md"
+    requirement = "The InterfaceKernel system shall use with interface material in 3D."
+                  "In this penalty implementation, jump is calculated by an interface material."
```
Comment (YaqiWang) [inline test/tests/interfacekernels/3d_interface/tests]: same

**Review by YaqiWang — COMMENTED** [review]
Just browsed through this PR. Looks good to me. It is missing `InternalSideMaterial` but I guess we can merge this first and add that later. If we do so, I suggest that we create a new issue for adding `InternalSideMaterial`.

**Comment by YaqiWang** [comment]
Thanks @lindsayad for making this happen. `InternalSideMaterial` is desired because it can be used to cleanup the penalty evaluations in Rattlesnake and remove several copies of DGKernels.

**Comment by lindsayad** [comment]
> Thanks @lindsayad for making this happen. `InternalSideMaterial` is desired because it can be used to cleanup the penalty evaluations in Rattlesnake and remove several copies of DGKernels.

You're welcome. Cool, yea I will make sure to work on `InternalSideMaterial` next. After creating this, it should be pretty easy to follow this design pattern and create `InternalSideMaterial`.

File: framework/include/materials/InterfaceMaterial.h [hunk 10]
```diff
@@ -0,0 +1,239 @@
...
+  const MaterialProperty<T> & getNeighborMaterialPropertyByName(const std::string & prop_name);
+  template <typename T>
+  const MaterialProperty<T> & getNeighborMaterialPropertyOldByName(const std::string & prop_name);
+  template <typename T>
+  const MaterialProperty<T> & getNeighborMaterialPropertyOlderByName(const std::string & prop_name);
+  ///@}
+
+  using MaterialBase::getZeroMaterialProperty;
+
+  virtual bool isBoundaryMaterial() const override { return _bnd; }
```
[the remark this answers is not in the document]
Comment (lindsayad) [reply framework/include/materials/InterfaceMaterial.h]: Indeed!

File: test/src/materials/InterfaceValueMaterial.C (hunk 3 above)
Comment (lindsayad) [reply test/src/materials/InterfaceValueMaterial.C]: This was initially a `mooseWarning` created by @arovinelli; I changed it to a `mooseAssert`. He developed this test to make sure that he could feed in a non-zero neighbor var and a neighbor material property and make sure that they match (with something like `MaterialRealAux` involved somewhere)

File: test/tests/interfacekernels/1d_interface/coupled_value_coupled_flux_with_jump_material.i (hunk 5 above)
Comment (lindsayad) [reply test/tests/interfacekernels/1d_interface/coupled_value_coupled_flux_with_jump_material.i]: It is a toy though; it's in the test directory 😄

**Comment by lindsayad** [comment]
I have a fix ready for rattlesnake

**Comment by lindsayad** [comment]
Thanks for your review @YaqiWang. I think I've addressed most things

File: framework/src/materials/InterfaceMaterial.C [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+  : MaterialBase(parameters),
+    NeighborCoupleable(this, false, false),
+    TwoMaterialPropertyInterface(this, blockIDs(), boundaryIDs()),
+    _q_point(_assembly.qPointsFace()),
+    _qrule(_assembly.qRuleFace()),
+    _JxW(_assembly.JxWFace()),
+    _current_elem(_assembly.elem()),
+    _neighbor_elem(_assembly.neighbor()),
+    _current_side(_assembly.side()),
+    _neighbor_side(_assembly.neighborSide())
```
Comment (arovinelli) [inline framework/src/materials/InterfaceMaterial.C]: you might want add the normal too,  but this is just a suggestion as I'don't know how much overhead there will be

File: test/src/materials/InterfaceValueMaterial.C (hunk 3 above)
Comment (arovinelli) [reply test/src/materials/InterfaceValueMaterial.C]: yep @lindsayad perfectly got the idea.
This was a check to make sure that using a material property or an aux variable would not affect the value of the jump. I agree with modifying it in an assert

**Review by arovinelli — COMMENTED** [review]
@lindsayad @permcody @YaqiWang This is a pretty big change to MOOSE I'm wiring this as a comment but should really be a requirement. 
I might be wrong but I don't see any test using `getNeighborMatarialPropertyOld` (for the interface material) or the interface kernel using `getMatarialPropertyOld` (an old interface material property).
I might be overnparanoid but it might be something to add.
BTW @lindsayad thanks for taking the time to implement this, I really appreciate it!

**Comment by arovinelli** [comment]
@lindsayad I check out you branch and added a few tests making sure the new material can couple old variable values and material properties. You can find this changes in my [branch](https://github.com/arovinelli/moose/tree/material_base_alex) 
These changes should not break any other test
Thanks again for working on this

**Comment by lindsayad** [comment]
Thanks @arovinelli I've incorporated those tests

**Comment by arovinelli** [comment]
@lindsayad I forgot that when using `BreakMeshByBlockGenerator` coupled with `InterfaceKernel` we cannot recover. Also when using `BreakMeshByBlockGenerator` the mesh can only be replicated.
I update the tests in a new commit [here](https://github.com/arovinelli/moose/tree/material_base_alex)

**Review by arovinelli — APPROVED** [review]
@lindsayad @permcody @YaqiWang 
For me this is good to go

**Comment by arovinelli** [comment]
@permcody @lindsayad @YaqiWang 
does this PR requires something else before being merged?

**Comment by lindsayad** [comment]
I’m just waiting on review

**Comment by permcody** [comment]
I'll review it.

**Review by permcody — APPROVED** [review]

**Comment by permcody** [comment]
@lindsayad thanks for pushing this over the finish line. I know it's been a long time coming and it was a lot of work. Excellent work. @arovinelli - thanks for the original contribution to get us down the down the path.

**Comment by arovinelli** [comment]
@permcody and @alex,  thanks a lot for making this happen. It was a long way but the interface material is now a reality,
I'll finally be able to switch to current MOOSE  and add some CZM model to tensormechanics, such that everyone can use them.

Cheers and thanks again!

All the best!!

> @lindsayad thanks for pushing this over the finish line. I know it's been a long time coming and it was a lot of work. Excellent work. @arovinelli - thanks for the original contribution to get us down the down the path.

**Pull request merged**